// @ngxf/socket.io/esm5/lib/directives/socket.io.on.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/socket.io.on.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { fromEvent } from 'rxjs';
import 'socket.io-client';
/**
 * @record
 */
function SocketIOOnContext() { }
if (false) {
    /** @type {?} */
    SocketIOOnContext.prototype.$implicit;
}
var SocketIOOnDirective = /** @class */ (function () {
    function SocketIOOnDirective(templateRef, vcr) {
        this.templateRef = templateRef;
        this.vcr = vcr;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    SocketIOOnDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.socketIOOnOn || changes.socketIOOnFrom) {
            this.destroy();
            if (this.socketIOOnOn && this.socketIOOnFrom) {
                if (changes.socketIOOnOn.previousValue !== this.socketIOOnOn ||
                    changes.socketIOOnFrom.previousValue !== this.socketIOOnFrom) {
                    this.init();
                }
            }
        }
    };
    /**
     * @return {?}
     */
    SocketIOOnDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    /**
     * @private
     * @return {?}
     */
    SocketIOOnDirective.prototype.init = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.subscription = fromEvent(this.socketIOOnFrom, this.socketIOOnOn)
            .subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.context.$implicit = value;
            _this.viewRef = _this.templateRef.createEmbeddedView(_this.context);
            _this.viewRef.detectChanges();
            if (_this.viewRef) {
                _this.viewRef.destroy();
                _this.viewRef = null;
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    SocketIOOnDirective.prototype.destroy = /**
     * @private
     * @return {?}
     */
    function () {
        this.vcr.clear();
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
    };
    SocketIOOnDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[socketIOOn]'
                },] }
    ];
    /** @nocollapse */
    SocketIOOnDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    SocketIOOnDirective.propDecorators = {
        socketIOOnOn: [{ type: Input }],
        socketIOOnFrom: [{ type: Input }]
    };
    return SocketIOOnDirective;
}());
export { SocketIOOnDirective };
if (false) {
    /** @type {?} */
    SocketIOOnDirective.prototype.socketIOOnOn;
    /** @type {?} */
    SocketIOOnDirective.prototype.socketIOOnFrom;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.vcr;
}
