// @ngxf/platform/esm5/lib/tools/recompose/composed.view.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tools/recompose/composed.view.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __read } from "tslib";
import { omit } from '../utils';
/** @type {?} */
export var PARENT_CONTEXT_TOKEN = Symbol('[PARENT_CONTEXT_TOKEN]');
/**
 * @record
 */
export function ComposedContext() { }
if (false) {
    /* Skipping unnamed member:
    [ PARENT_CONTEXT_TOKEN ]: TemplateRef<ComposedContext>[];*/
    /** @type {?} */
    ComposedContext.prototype.children;
    /* Skipping unhandled member: [ key: string ]: any;*/
}
/**
 * @template C
 */
var /**
 * @template C
 */
ComposedView = /** @class */ (function () {
    function ComposedView(viewContainerRef, templateRefs) {
        this.viewContainerRef = viewContainerRef;
        this.templateRefs = templateRefs;
    }
    /**
     * @param {?} context
     * @return {?}
     */
    ComposedView.prototype.createEmbeddedView = /**
     * @param {?} context
     * @return {?}
     */
    function (context) {
        var _a = __read(this.getParentTemplateRefs()), templateRef = _a[0], tail = _a.slice(1);
        this.context = (/** @type {?} */ ((/** @type {?} */ (context))));
        if (tail.length > 0) {
            this.elementRef = templateRef.elementRef;
            this.context[PARENT_CONTEXT_TOKEN] = tail;
            this.context.children = new ComposedView(this.viewContainerRef, tail);
        }
        this.viewRef =
            this.viewContainerRef.createEmbeddedView(templateRef, this.context);
        return (/** @type {?} */ ((/** @type {?} */ (this.viewRef))));
    };
    /**
     * @return {?}
     */
    ComposedView.prototype.markForCheck = /**
     * @return {?}
     */
    function () {
        if (this.viewRef) {
            this.viewRef.markForCheck();
        }
    };
    /**
     * @return {?}
     */
    ComposedView.prototype.detectChanges = /**
     * @return {?}
     */
    function () {
        if (this.viewRef) {
            this.viewRef.detectChanges();
        }
    };
    /**
     * @param {?} update
     * @return {?}
     */
    ComposedView.prototype.updateContext = /**
     * @param {?} update
     * @return {?}
     */
    function (update) {
        /** @type {?} */
        var parent = this.getHydratedParentContext();
        update(this.context, parent);
        this.markForCheck();
    };
    /**
     * @return {?}
     */
    ComposedView.prototype.destroy = /**
     * @return {?}
     */
    function () {
        if (this.viewRef && !this.viewRef.destroyed) {
            this.viewRef.destroy();
        }
        this.viewRef = null;
    };
    /**
     * @private
     * @return {?}
     */
    ComposedView.prototype.getParentTemplateRefs = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.templateRefs) {
            return this.templateRefs;
        }
        /** @type {?} */
        var view = ((/** @type {?} */ (this.viewContainerRef.injector))).view;
        /** @type {?} */
        var context = findParentContext(view);
        return context[PARENT_CONTEXT_TOKEN];
    };
    /**
     * @private
     * @return {?}
     */
    ComposedView.prototype.getHydratedParentContext = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var view = ((/** @type {?} */ (this.viewContainerRef.injector))).view;
        return omit(findParentContext(view), [PARENT_CONTEXT_TOKEN, 'children']);
    };
    return ComposedView;
}());
/**
 * @template C
 */
export { ComposedView };
if (false) {
    /** @type {?} */
    ComposedView.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.templateRefs;
}
/**
 * @param {?} view
 * @return {?}
 */
export function findParentContext(view) {
    /** @type {?} */
    var context = view.context;
    if (isParentContext(context)) {
        return context;
    }
    return findParentContext(view.parent);
}
/**
 * @param {?} context
 * @return {?}
 */
export function isParentContext(context) {
    return context && PARENT_CONTEXT_TOKEN in context;
}
