// @ngxf/platform/esm5/lib/tools/virtual.handler.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tools/virtual.handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var VirtualHandler = /** @class */ (function () {
    function VirtualHandler() {
        var _this = this;
        this.elementLink = new Map();
        this.observer = new IntersectionObserver((/**
         * @param {?} entries
         * @return {?}
         */
        function (entries) {
            entries.forEach((/**
             * @param {?} entry
             * @return {?}
             */
            function (entry) {
                /** @type {?} */
                var cd = _this.elementLink.get((/** @type {?} */ (entry.target)));
                entry.isIntersecting ? attachCD(cd) : detachCD(cd);
            }));
        }));
    }
    /**
     * @param {?} element
     * @param {?} cd
     * @return {?}
     */
    VirtualHandler.prototype.register = /**
     * @param {?} element
     * @param {?} cd
     * @return {?}
     */
    function (element, cd) {
        this.elementLink.set(element, cd);
        this.observer.observe(element);
    };
    /**
     * @param {?} element
     * @return {?}
     */
    VirtualHandler.prototype.unregister = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        this.elementLink.delete(element);
        this.observer.unobserve(element);
    };
    /**
     * @return {?}
     */
    VirtualHandler.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.elementLink.forEach((/**
         * @param {?} cd
         * @param {?} e
         * @return {?}
         */
        function (cd, e) { return _this.observer.unobserve(e); }));
        this.elementLink.clear();
    };
    VirtualHandler.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ VirtualHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function VirtualHandler_Factory() { return new VirtualHandler(); }, token: VirtualHandler, providedIn: "root" });
    return VirtualHandler;
}());
export { VirtualHandler };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VirtualHandler.prototype.elementLink;
    /**
     * @type {?}
     * @private
     */
    VirtualHandler.prototype.observer;
}
/**
 * @param {?} cd
 * @return {?}
 */
function attachCD(cd) {
    cd.reattach();
    cd.detectChanges();
}
/**
 * @param {?} cd
 * @return {?}
 */
function detachCD(cd) {
    cd.detach();
}
