// @datorama/akita-ng-router-store/__ivy_ngcc__/fesm2015/datorama-akita-ng-router-store.js
import { ɵɵdefineInjectable, Injectable, ɵɵinject, NgModule } from '@angular/core';
import { __decorate, __metadata } from 'tslib';
import { RoutesRecognized, GuardsCheckEnd, ResolveEnd, NavigationCancel, NavigationError, NavigationEnd, Router } from '@angular/router';
import { Store, StoreConfig, filterNil, Query, setSkipAction, action } from '@datorama/akita';
import { Subject, combineLatest } from 'rxjs';
import { map, distinctUntilChanged, pluck } from 'rxjs/operators';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
function createInitialRouterState() {
    return {
        state: null,
        navigationId: null
    };
}
let RouterStore = class RouterStore extends Store {
    constructor() {
        super(createInitialRouterState());
    }
};
RouterStore.ɵfac = function RouterStore_Factory(t) { return new (t || RouterStore)(); };
RouterStore.ɵprov = ɵɵdefineInjectable({ factory: function RouterStore_Factory() { return new RouterStore(); }, token: RouterStore, providedIn: "root" });
RouterStore.ctorParameters = () => [];
RouterStore = __decorate([
    StoreConfig({ name: 'router' }),
    __metadata("design:paramtypes", [])
], RouterStore);

function slice(section) {
    return (source) => {
        return source.pipe(map((data) => data.state)).pipe(filterNil, map((state) => state[section]));
    };
}
class RouterQuery extends Query {
    constructor(store) {
        super(store);
        this.store = store;
        this.__navigationCancel = new Subject();
        this.__navigationError = new Subject();
    }
    selectParams(names) {
        if (names === undefined) {
            return this.select().pipe(slice('params'), distinctUntilChanged());
        }
        const select = (p) => this.select().pipe(slice('params'), pluck(p), distinctUntilChanged());
        if (Array.isArray(names)) {
            const sources = names.map(select);
            return combineLatest(sources);
        }
        return select(names).pipe(distinctUntilChanged());
    }
    getParams(name) {
        if (this.getValue().state) {
            const params = this.getValue().state.params;
            if (name === undefined) {
                return params;
            }
            return params[name];
        }
        return null;
    }
    selectQueryParams(names) {
        if (names === undefined) {
            return this.select().pipe(slice('queryParams'), distinctUntilChanged());
        }
        const select = (p) => this.select().pipe(slice('queryParams'), pluck(p), distinctUntilChanged());
        if (Array.isArray(names)) {
            const sources = names.map(select);
            return combineLatest(sources);
        }
        return select(names);
    }
    getQueryParams(name) {
        if (this.getValue().state) {
            const params = this.getValue().state.queryParams;
            if (name === undefined) {
                return params;
            }
            return params[name];
        }
        return null;
    }
    selectFragment() {
        return this.select().pipe(slice('fragment'), distinctUntilChanged());
    }
    getFragment() {
        if (this.getValue().state) {
            return this.getValue().state.fragment;
        }
        return null;
    }
    selectData(name) {
        if (name === undefined) {
            return this.select().pipe(slice('data'), distinctUntilChanged());
        }
        return this.select().pipe(slice('data'), pluck(name), distinctUntilChanged());
    }
    getData(name) {
        if (this.getValue().state) {
            const data = this.getValue().state.data;
            if (name === undefined) {
                return data;
            }
            return data[name];
        }
        return null;
    }
    selectNavigationExtras(name) {
        if (name === undefined) {
            return this.select().pipe(slice('navigationExtras'), distinctUntilChanged());
        }
        return this.select().pipe(slice('data'), pluck(name), distinctUntilChanged());
    }
    getNavigationExtras(name) {
        if (this.getValue().state) {
            const data = this.getValue().state.navigationExtras;
            if (name === undefined) {
                return data;
            }
            return data[name];
        }
        return null;
    }
    selectNavigationCancel() {
        return this.__navigationCancel.asObservable();
    }
    selectNavigationError() {
        return this.__navigationError.asObservable();
    }
}
RouterQuery.ɵfac = function RouterQuery_Factory(t) { return new (t || RouterQuery)(ɵngcc0.ɵɵinject(RouterStore)); };
RouterQuery.ctorParameters = () => [
    { type: RouterStore }
];
RouterQuery.ɵprov = ɵɵdefineInjectable({ factory: function RouterQuery_Factory() { return new RouterQuery(ɵɵinject(RouterStore)); }, token: RouterQuery, providedIn: "root" });
RouterQuery.ctorParameters = () => [
    { type: RouterStore }
];

class RouterService {
    constructor(routerStore, routerQuery, router) {
        this.routerStore = routerStore;
        this.routerQuery = routerQuery;
        this.router = router;
        this.dispatchTriggeredByRouter = false;
        this.navigationTriggeredByDispatch = false;
    }
    dispatchRouterCancel(event) {
        this.update({ navigationId: event.id });
        this.routerQuery.__navigationCancel.next(event);
    }
    dispatchRouterError(event) {
        this.update({ navigationId: event.id });
        this.routerQuery.__navigationError.next(event);
    }
    dispatchRouterSuccess() {
        this.update(this.lastRouterState);
    }
    init() {
        this.setUpStoreListener();
        this.setUpStateRollbackEvents();
    }
    update(routerState) {
        this.dispatchTriggeredByRouter = true;
        this.routerStore.update((state) => {
            return Object.assign(Object.assign({}, state), routerState);
        });
        this.dispatchTriggeredByRouter = false;
        this.navigationTriggeredByDispatch = false;
    }
    setUpStoreListener() {
        this.routerQuery
            .select((state) => state)
            .subscribe((s) => {
            this.lastRouterState = s;
            this.navigateIfNeeded();
        });
    }
    navigateIfNeeded() {
        if (!this.lastRouterState || !this.lastRouterState.state || this.dispatchTriggeredByRouter) {
            return;
        }
        if (this.router.url !== this.lastRouterState.state.url) {
            this.navigationTriggeredByDispatch = true;
            setSkipAction();
            this.router.navigateByUrl(this.lastRouterState.state.url);
        }
    }
    setUpStateRollbackEvents() {
        this.router.events.subscribe((e) => {
            if (e instanceof RoutesRecognized || e instanceof GuardsCheckEnd || e instanceof ResolveEnd) {
                this.lastRouterState = this.serializeRoute(e);
            }
            else if (e instanceof NavigationCancel) {
                this.dispatchRouterCancel(e);
            }
            else if (e instanceof NavigationError) {
                this.dispatchRouterError(e);
            }
            else if (e instanceof NavigationEnd && !this.navigationTriggeredByDispatch) {
                this.dispatchRouterSuccess();
            }
        });
    }
    serializeRoute(navigationEvent) {
        let state = navigationEvent.state.root;
        while (state.firstChild) {
            state = state.firstChild;
        }
        const { params, data, queryParams, fragment } = state;
        return {
            navigationId: navigationEvent.id,
            state: {
                url: navigationEvent.url,
                urlAfterRedirects: navigationEvent.urlAfterRedirects,
                params,
                queryParams,
                fragment,
                data,
                navigationExtras: this.router.getCurrentNavigation().extras ? this.router.getCurrentNavigation().extras.state : {},
            },
        };
    }
}
RouterService.ɵfac = function RouterService_Factory(t) { return new (t || RouterService)(ɵngcc0.ɵɵinject(RouterStore), ɵngcc0.ɵɵinject(RouterQuery), ɵngcc0.ɵɵinject(ɵngcc1.Router)); };
RouterService.ctorParameters = () => [
    { type: RouterStore },
    { type: RouterQuery },
    { type: Router }
];
RouterService.ɵprov = ɵɵdefineInjectable({ factory: function RouterService_Factory() { return new RouterService(ɵɵinject(RouterStore), ɵɵinject(RouterQuery), ɵɵinject(Router)); }, token: RouterService, providedIn: "root" });
RouterService.ctorParameters = () => [
    { type: RouterStore },
    { type: RouterQuery },
    { type: Router }
];
__decorate([
    action('Navigation Cancelled'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [NavigationCancel]),
    __metadata("design:returntype", void 0)
], RouterService.prototype, "dispatchRouterCancel", null);
__decorate([
    action('Navigation Error'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [NavigationError]),
    __metadata("design:returntype", void 0)
], RouterService.prototype, "dispatchRouterError", null);
__decorate([
    action('Navigation Succeeded'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], RouterService.prototype, "dispatchRouterSuccess", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RouterStore, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RouterQuery, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: RouterStore }]; }, null); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RouterService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: RouterStore }, { type: RouterQuery }, { type: ɵngcc1.Router }]; }, null); })();

class AkitaNgRouterStoreModule {
    constructor(routerService) {
        this.routerService = routerService;
        this.routerService.init();
    }
}
AkitaNgRouterStoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AkitaNgRouterStoreModule });
AkitaNgRouterStoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AkitaNgRouterStoreModule_Factory(t) { return new (t || AkitaNgRouterStoreModule)(ɵngcc0.ɵɵinject(RouterService)); } });
AkitaNgRouterStoreModule.ctorParameters = () => [
    { type: RouterService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AkitaNgRouterStoreModule, [{
        type: NgModule
    }], function () { return [{ type: RouterService }]; }, null); })();
AkitaNgRouterStoreModule.ctorParameters = () => [
    { type: RouterService }
];

/**
 * Generated bundle index. Do not edit.
 */

export { AkitaNgRouterStoreModule, RouterQuery, RouterService, RouterStore, createInitialRouterState };
