// @ngxf/platform/esm5/lib/pipes/console.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/console.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __read, __spread } from "tslib";
import { Inject, InjectionToken, Pipe } from '@angular/core';
/**
 * @return {?}
 */
export function consoleFactory() {
    return console;
}
/** @type {?} */
export var CONSOLE = new InjectionToken('Console', {
    providedIn: 'root',
    factory: consoleFactory
});
/** @type {?} */
export var skipValueOperators = [
    'count',
    'markTimeline',
    'time',
    'timeEnd',
    'profile',
    'profileEnd',
    'timeline',
    'timelineEnd',
    'timeStamp',
    'group',
    'groupCollapsed'
];
var ConsolePipe = /** @class */ (function () {
    function ConsolePipe(console) {
        this.console = console;
    }
    /**
     * @template T
     * @param {?} message
     * @param {?=} logLevel
     * @param {...?} optionalParams
     * @return {?}
     */
    ConsolePipe.prototype.transform = /**
     * @template T
     * @param {?} message
     * @param {?=} logLevel
     * @param {...?} optionalParams
     * @return {?}
     */
    function (message, logLevel) {
        var _a;
        if (logLevel === void 0) { logLevel = 'log'; }
        var optionalParams = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            optionalParams[_i - 2] = arguments[_i];
        }
        /** @type {?} */
        var params;
        if (skipValueOperators.includes(logLevel)) {
            if (optionalParams.length > 0) {
                params = optionalParams;
            }
            else {
                params = [message];
            }
        }
        else {
            params = __spread([message], optionalParams);
        }
        (_a = this.console)[logLevel].apply(_a, __spread(params));
        return message;
    };
    ConsolePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'console'
                },] }
    ];
    /** @nocollapse */
    ConsolePipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CONSOLE,] }] }
    ]; };
    return ConsolePipe;
}());
export { ConsolePipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConsolePipe.prototype.console;
}
