// @ngxf/platform/esm2015/ngxf-platform.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngxf-platform.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxfModule, AsyncDirective, ComposeDirective, CookiesDirective, HttpDirective, InitDirective, IsDevModeDirective, KeepAliveDirective, KeepAliveConfig, KeepAliveGC, LazyDirective, LAZY_COMPONENT_TOKEN, NestDirective, RenamePropDirective, RepeatDirective, ReturnDirective, RouteDirective, SetPropsDirective, TimeoutDirective, NgForTrackByKeyDirective, UseEffectDirective, UseReducerDirective, UseStateDirective, VirtualDirective, CallPipe, ConsolePipe, UseEffect, UseReducer, UseState } from './public_api';
export { DevModeService as ɵa } from './lib/directives/is-dev-mode.directive';
export { KeepAliveManager as ɵc, KeepAliveViewRefs as ɵb } from './lib/directives/keep-alive.directive';
export { RepeatContext as ɵd } from './lib/directives/repeat.directive';
export { CONSOLE as ɵf, consoleFactory as ɵe } from './lib/pipes/console.pipe';
export { CookieOptions as ɵg, CookiesService as ɵh } from './lib/tools/cookies.tools';
export { VirtualHandler as ɵi } from './lib/tools/virtual.handler';
