// @ngxf/platform/esm2015/lib/pipes/console.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/console.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, InjectionToken, Pipe } from '@angular/core';
/**
 * @return {?}
 */
export function consoleFactory() {
    return console;
}
/** @type {?} */
export const CONSOLE = new InjectionToken('Console', {
    providedIn: 'root',
    factory: consoleFactory
});
/** @type {?} */
export const skipValueOperators = [
    'count',
    'markTimeline',
    'time',
    'timeEnd',
    'profile',
    'profileEnd',
    'timeline',
    'timelineEnd',
    'timeStamp',
    'group',
    'groupCollapsed'
];
export class ConsolePipe {
    /**
     * @param {?} console
     */
    constructor(console) {
        this.console = console;
    }
    /**
     * @template T
     * @param {?} message
     * @param {?=} logLevel
     * @param {...?} optionalParams
     * @return {?}
     */
    transform(message, logLevel = 'log', ...optionalParams) {
        /** @type {?} */
        let params;
        if (skipValueOperators.includes(logLevel)) {
            if (optionalParams.length > 0) {
                params = optionalParams;
            }
            else {
                params = [message];
            }
        }
        else {
            params = [message, ...optionalParams];
        }
        this.console[logLevel](...params);
        return message;
    }
}
ConsolePipe.decorators = [
    { type: Pipe, args: [{
                name: 'console'
            },] }
];
/** @nocollapse */
ConsolePipe.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CONSOLE,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConsolePipe.prototype.console;
}
