// @ngxf/platform/esm5/lib/directives/compose.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/compose.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __read, __spread } from "tslib";
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { ComposedView } from '../tools/recompose/composed.view';
/**
 * @record
 */
function ComposeContext() { }
if (false) {
    /** @type {?} */
    ComposeContext.prototype.$implicit;
    /** @type {?} */
    ComposeContext.prototype.compose;
}
var ComposeDirective = /** @class */ (function () {
    function ComposeDirective(templateRef, viewContainerRef) {
        this.templateRef = templateRef;
        this.viewContainerRef = viewContainerRef;
        this.context = { $implicit: null, compose: null };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ComposeDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('compose' in changes) {
            this.onTemplateRefsDidChanged(this.compose, changes.compose.previousValue);
        }
        if ('composeOf' in changes) {
            this.onTemplateRefsDidChanged(this.composeOf, changes.composeOf.previousValue);
        }
    };
    /**
     * @return {?}
     */
    ComposeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    /**
     * @private
     * @param {?} current
     * @param {?} previous
     * @return {?}
     */
    ComposeDirective.prototype.onTemplateRefsDidChanged = /**
     * @private
     * @param {?} current
     * @param {?} previous
     * @return {?}
     */
    function (current, previous) {
        if (!this.viewRef) {
            return this.create(current);
        }
        if (this.viewRef.destroyed || current !== previous) {
            this.destroy();
            return this.onTemplateRefsDidChanged(current, null);
        }
    };
    /**
     * @private
     * @param {?} templateRefs
     * @return {?}
     */
    ComposeDirective.prototype.create = /**
     * @private
     * @param {?} templateRefs
     * @return {?}
     */
    function (templateRefs) {
        this.context.compose = this.context.$implicit = this.createComposeFn(templateRefs);
        this.viewRef =
            this.viewContainerRef.createEmbeddedView(this.templateRef, this.context);
    };
    /**
     * @private
     * @return {?}
     */
    ComposeDirective.prototype.destroy = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.viewRef && !this.viewRef.destroyed) {
            this.viewRef.destroy();
        }
        this.viewRef = null;
    };
    /**
     * @private
     * @param {?} templateRefs
     * @return {?}
     */
    ComposeDirective.prototype.createComposeFn = /**
     * @private
     * @param {?} templateRefs
     * @return {?}
     */
    function (templateRefs) {
        var _this = this;
        return (/**
         * @param {?} templateRef
         * @return {?}
         */
        function (templateRef) {
            return new ComposedView(_this.viewContainerRef, __spread(templateRefs, [templateRef]));
        });
    };
    ComposeDirective.decorators = [
        { type: Directive, args: [{ selector: '[compose]' },] }
    ];
    /** @nocollapse */
    ComposeDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    ComposeDirective.propDecorators = {
        compose: [{ type: Input }],
        composeOf: [{ type: Input }]
    };
    return ComposeDirective;
}());
export { ComposeDirective };
if (false) {
    /** @type {?} */
    ComposeDirective.prototype.compose;
    /** @type {?} */
    ComposeDirective.prototype.composeOf;
    /**
     * @type {?}
     * @private
     */
    ComposeDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ComposeDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    ComposeDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    ComposeDirective.prototype.viewContainerRef;
}
