// @ngxf/platform/esm5/lib/directives/is-dev-mode.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/is-dev-mode.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef, ViewContainerRef, isDevMode, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var DevModeService = /** @class */ (function () {
    function DevModeService() {
    }
    /**
     * @return {?}
     */
    DevModeService.prototype.isDevMode = /**
     * @return {?}
     */
    function () {
        return isDevMode();
    };
    DevModeService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    DevModeService.ctorParameters = function () { return []; };
    /** @nocollapse */ DevModeService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DevModeService_Factory() { return new DevModeService(); }, token: DevModeService, providedIn: "root" });
    return DevModeService;
}());
export { DevModeService };
var IsDevModeDirective = /** @class */ (function () {
    function IsDevModeDirective(devModeService, viewContainerRef, templateRef) {
        if (devModeService.isDevMode()) {
            viewContainerRef.createEmbeddedView(templateRef);
        }
    }
    /**
     * @return {?}
     */
    IsDevModeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
    };
    IsDevModeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[isDevMode]'
                },] }
    ];
    /** @nocollapse */
    IsDevModeDirective.ctorParameters = function () { return [
        { type: DevModeService },
        { type: ViewContainerRef },
        { type: TemplateRef }
    ]; };
    return IsDevModeDirective;
}());
export { IsDevModeDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IsDevModeDirective.prototype.viewRef;
}
