// @ngxf/platform/esm5/lib/directives/rename-prop.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/rename-prop.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { ComposedView } from '../tools/recompose/composed.view';
import { omit } from '../tools/utils/index';
/**
 * @record
 */
function RenamePropContext() { }
if (false) {
    /** @type {?} */
    RenamePropContext.prototype.$implicit;
}
var RenamePropDirective = /** @class */ (function () {
    function RenamePropDirective(templateRef, viewContainerRef) {
        this.templateRef = templateRef;
        this.viewContainerRef = viewContainerRef;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    RenamePropDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('renameProp' in changes && 'renamePropTo' in changes) {
            this.onRenameDidChanged(this.renameProp, this.renamePropTo, changes.renameProp.previousValue, changes.renamePropTo.previousValue);
        }
    };
    /**
     * @return {?}
     */
    RenamePropDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    /**
     * @private
     * @param {?} from
     * @param {?} to
     * @param {?} fromPrevious
     * @param {?} toPrevious
     * @return {?}
     */
    RenamePropDirective.prototype.onRenameDidChanged = /**
     * @private
     * @param {?} from
     * @param {?} to
     * @param {?} fromPrevious
     * @param {?} toPrevious
     * @return {?}
     */
    function (from, to, fromPrevious, toPrevious) {
        if (!this.composedView) {
            this.create();
        }
        if (from !== fromPrevious || to !== toPrevious) {
            this.rename(from, to);
        }
    };
    /**
     * @private
     * @return {?}
     */
    RenamePropDirective.prototype.create = /**
     * @private
     * @return {?}
     */
    function () {
        this.composedView = new ComposedView(this.viewContainerRef);
        this.composedView.createEmbeddedView(this.context);
    };
    /**
     * @private
     * @param {?} from
     * @param {?} to
     * @return {?}
     */
    RenamePropDirective.prototype.rename = /**
     * @private
     * @param {?} from
     * @param {?} to
     * @return {?}
     */
    function (from, to) {
        this.composedView.updateContext((/**
         * @param {?} context
         * @param {?} parent
         * @return {?}
         */
        function (context, parent) {
            var _a;
            return Object.assign(context, omit(parent, [from]), (_a = {}, _a[to] = parent[from], _a));
        }));
    };
    /**
     * @private
     * @return {?}
     */
    RenamePropDirective.prototype.destroy = /**
     * @private
     * @return {?}
     */
    function () {
        this.viewContainerRef.clear();
        if (this.composedView) {
            this.composedView.destroy();
            this.composedView = null;
        }
    };
    RenamePropDirective.decorators = [
        { type: Directive, args: [{ selector: '[renameProp]' },] }
    ];
    /** @nocollapse */
    RenamePropDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    RenamePropDirective.propDecorators = {
        renameProp: [{ type: Input }],
        renamePropTo: [{ type: Input }]
    };
    return RenamePropDirective;
}());
export { RenamePropDirective };
if (false) {
    /** @type {?} */
    RenamePropDirective.prototype.renameProp;
    /** @type {?} */
    RenamePropDirective.prototype.renamePropTo;
    /**
     * @type {?}
     * @private
     */
    RenamePropDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    RenamePropDirective.prototype.composedView;
    /**
     * @type {?}
     * @private
     */
    RenamePropDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    RenamePropDirective.prototype.viewContainerRef;
}
