// @ngxf/platform/esm5/lib/directives/repeat.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/repeat.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
var RepeatContext = /** @class */ (function () {
    function RepeatContext() {
    }
    return RepeatContext;
}());
export { RepeatContext };
if (false) {
    /** @type {?} */
    RepeatContext.prototype.$implicit;
    /** @type {?} */
    RepeatContext.prototype.repeat;
    /** @type {?} */
    RepeatContext.prototype.index;
}
var RepeatDirective = /** @class */ (function () {
    function RepeatDirective(viewContainerRef, templateRef) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
    }
    Object.defineProperty(RepeatDirective.prototype, "count", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var count = isNumber(this.repeat) ? this.repeat :
                isNumber(this.repeatOf) ? this.repeatOf : 0;
            return Math.max(count, 0);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    RepeatDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('repeat' in changes || 'repeatOf' in changes) {
            this.onRepeatChanged(this.count);
        }
    };
    /**
     * @return {?}
     */
    RepeatDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.viewContainerRef.clear();
    };
    /**
     * @private
     * @param {?} count
     * @return {?}
     */
    RepeatDirective.prototype.onRepeatChanged = /**
     * @private
     * @param {?} count
     * @return {?}
     */
    function (count) {
        /** When we need to create new items */
        for (var i = this.viewContainerRef.length; i < count; i++) {
            this.viewContainerRef.createEmbeddedView(this.templateRef, {
                $implicit: i, index: i, repeat: i
            }, i);
        }
        /** When we need to remove old items */
        for (var i = this.viewContainerRef.length; i > count; i--) {
            this.viewContainerRef.remove(i - 1);
        }
    };
    RepeatDirective.decorators = [
        { type: Directive, args: [{ selector: '[repeat]' },] }
    ];
    /** @nocollapse */
    RepeatDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: TemplateRef }
    ]; };
    RepeatDirective.propDecorators = {
        repeat: [{ type: Input }],
        repeatOf: [{ type: Input }]
    };
    return RepeatDirective;
}());
export { RepeatDirective };
if (false) {
    /** @type {?} */
    RepeatDirective.prototype.repeat;
    /** @type {?} */
    RepeatDirective.prototype.repeatOf;
    /**
     * @type {?}
     * @private
     */
    RepeatDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    RepeatDirective.prototype.templateRef;
}
/**
 * @param {?} value
 * @return {?}
 */
function isNumber(value) {
    return typeof value === 'number';
}
