// @ngxf/platform/esm5/lib/directives/return.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/return.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef } from '@angular/core';
import { ComposedView } from '../tools/recompose/composed.view';
/**
 * @record
 */
function ReturnContext() { }
if (false) {
    /** @type {?} */
    ReturnContext.prototype.$implicit;
}
var ReturnDirective = /** @class */ (function () {
    function ReturnDirective(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ReturnDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('return' in changes) {
            if (!this.composedView) {
                this.create();
            }
            else {
                this.update();
            }
        }
    };
    /**
     * @return {?}
     */
    ReturnDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.composedView) {
            this.composedView.destroy();
            this.composedView = null;
        }
    };
    /**
     * @private
     * @return {?}
     */
    ReturnDirective.prototype.create = /**
     * @private
     * @return {?}
     */
    function () {
        this.composedView = new ComposedView(this.viewContainerRef);
        this.composedView.createEmbeddedView(this.context);
        this.update();
    };
    /**
     * @private
     * @return {?}
     */
    ReturnDirective.prototype.update = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.composedView.updateContext((/**
         * @param {?} context
         * @return {?}
         */
        function (context) {
            return Object.assign(context, _this.return);
        }));
    };
    ReturnDirective.decorators = [
        { type: Directive, args: [{ selector: '[return]' },] }
    ];
    /** @nocollapse */
    ReturnDirective.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    ReturnDirective.propDecorators = {
        return: [{ type: Input }]
    };
    return ReturnDirective;
}());
export { ReturnDirective };
if (false) {
    /** @type {?} */
    ReturnDirective.prototype.return;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.composedView;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.viewContainerRef;
}
