// @ngxf/platform/esm5/lib/directives/timeout.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/timeout.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
var TimeoutDirective = /** @class */ (function () {
    function TimeoutDirective(templateRef, viewContainerRef) {
        this.templateRef = templateRef;
        this.viewContainerRef = viewContainerRef;
    }
    Object.defineProperty(TimeoutDirective.prototype, "timeout", {
        set: /**
         * @param {?} milliseconds
         * @return {?}
         */
        function (milliseconds) {
            this.dispose();
            if (isNumber(milliseconds)) {
                this.create(milliseconds);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TimeoutDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.dispose();
        this.viewContainerRef.clear();
    };
    /**
     * @private
     * @param {?} milliseconds
     * @return {?}
     */
    TimeoutDirective.prototype.create = /**
     * @private
     * @param {?} milliseconds
     * @return {?}
     */
    function (milliseconds) {
        var _this = this;
        this.timeoutId = setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.viewContainerRef) {
                _this.viewContainerRef.createEmbeddedView(_this.templateRef);
            }
        }), milliseconds);
    };
    /**
     * @private
     * @return {?}
     */
    TimeoutDirective.prototype.dispose = /**
     * @private
     * @return {?}
     */
    function () {
        this.viewContainerRef.clear();
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
    };
    TimeoutDirective.decorators = [
        { type: Directive, args: [{ selector: '[timeout]' },] }
    ];
    /** @nocollapse */
    TimeoutDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    TimeoutDirective.propDecorators = {
        timeout: [{ type: Input }]
    };
    return TimeoutDirective;
}());
export { TimeoutDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.timeoutId;
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.viewContainerRef;
}
/**
 * @param {?} value
 * @return {?}
 */
function isNumber(value) {
    return typeof value === 'number';
}
