// @ngxf/platform/esm5/lib/directives/track-by-key.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/track-by-key.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgForOf } from '@angular/common';
import { Directive, Host, Input, Optional, ViewContainerRef } from '@angular/core';
/**
 * @template T
 */
var NgForTrackByKeyDirective = /** @class */ (function () {
    function NgForTrackByKeyDirective(ngFor, viewContainerRef) {
        this.ngFor = ngFor;
        this.viewContainerRef = viewContainerRef;
        if (!ngFor) {
            throw new Error('TrackByKey should use with *ngFor!');
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    NgForTrackByKeyDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('ngForTrackByKey' in changes) {
            /** @type {?} */
            var key_1 = this.ngForTrackByKey;
            if (key_1) {
                this.ngFor.ngForTrackBy = (/**
                 * @param {?} index
                 * @param {?} item
                 * @return {?}
                 */
                function (index, item) { return item[key_1]; });
            }
            else {
                this.ngFor.ngForTrackBy = undefined;
            }
            this.ngFor['_differ'] = null;
            this.ngFor['_ngForOfDirty'] = true;
            this.viewContainerRef.clear();
            this.ngFor.ngDoCheck();
        }
    };
    NgForTrackByKeyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngForTrackByKey]'
                },] }
    ];
    /** @nocollapse */
    NgForTrackByKeyDirective.ctorParameters = function () { return [
        { type: NgForOf, decorators: [{ type: Host }, { type: Optional }] },
        { type: ViewContainerRef }
    ]; };
    NgForTrackByKeyDirective.propDecorators = {
        ngForTrackByKey: [{ type: Input }]
    };
    return NgForTrackByKeyDirective;
}());
export { NgForTrackByKeyDirective };
if (false) {
    /** @type {?} */
    NgForTrackByKeyDirective.prototype.ngForTrackByKey;
    /**
     * @type {?}
     * @private
     */
    NgForTrackByKeyDirective.prototype.ngFor;
    /**
     * @type {?}
     * @private
     */
    NgForTrackByKeyDirective.prototype.viewContainerRef;
}
