// @ngxf/platform/esm5/lib/directives/use-effect.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/use-effect.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
var UseEffectDirective = /** @class */ (function () {
    function UseEffectDirective(templateRef, vcr) {
        this.templateRef = templateRef;
        this.vcr = vcr;
        this.context = {};
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    UseEffectDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.useEffect && this.useEffect) {
            this.destroy();
            this.init();
        }
        else if (changes.useEffectOn) {
            if (Array.isArray(this.useEffectOn)) {
                /** @type {?} */
                var previous = changes.useEffectOn.previousValue || [];
                /** @type {?} */
                var current_1 = changes.useEffectOn.currentValue || [];
                /** @type {?} */
                var isChanged = !previous.every((/**
                 * @param {?} item
                 * @param {?} index
                 * @return {?}
                 */
                function (item, index) { return current_1[index] === item; }));
                if (isChanged) {
                    this.destroy();
                    this.init();
                }
            }
            else {
                this.destroy();
                this.init();
            }
        }
    };
    /**
     * @return {?}
     */
    UseEffectDirective.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        if (!Array.isArray(this.useEffectOn)) {
            this.destroy();
            this.init();
        }
    };
    /**
     * @return {?}
     */
    UseEffectDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    /**
     * @private
     * @return {?}
     */
    UseEffectDirective.prototype.init = /**
     * @private
     * @return {?}
     */
    function () {
        this.viewRef = this.vcr.createEmbeddedView(this.templateRef, this.context);
        if (this.useEffect) {
            this.onDestroyCallback = this.useEffect();
        }
    };
    /**
     * @private
     * @return {?}
     */
    UseEffectDirective.prototype.destroy = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.onDestroyCallback) {
            this.onDestroyCallback();
            this.onDestroyCallback = null;
        }
        this.vcr.clear();
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
    };
    UseEffectDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[useEffect]'
                },] }
    ];
    /** @nocollapse */
    UseEffectDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    UseEffectDirective.propDecorators = {
        useEffect: [{ type: Input }],
        useEffectOn: [{ type: Input }]
    };
    return UseEffectDirective;
}());
export { UseEffectDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UseEffectDirective.prototype.context;
    /** @type {?} */
    UseEffectDirective.prototype.useEffect;
    /** @type {?} */
    UseEffectDirective.prototype.useEffectOn;
    /**
     * @type {?}
     * @private
     */
    UseEffectDirective.prototype.onDestroyCallback;
    /**
     * @type {?}
     * @private
     */
    UseEffectDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    UseEffectDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    UseEffectDirective.prototype.vcr;
}
