// @ngxf/platform/esm5/lib/directives/use-state.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/use-state.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
/**
 * @record
 * @template T
 */
function UseStateContext() { }
if (false) {
    /** @type {?} */
    UseStateContext.prototype.$implicit;
}
var UseStateDirective = /** @class */ (function () {
    function UseStateDirective(templateRef, vcr) {
        var _this = this;
        this.templateRef = templateRef;
        this.vcr = vcr;
        this.context = (/** @type {?} */ ({}));
        Object.defineProperty(this.context, '$implicit', {
            get: (/**
             * @return {?}
             */
            function () { return _this.value; }),
            set: (/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                _this.value = value;
                if (_this.embeddedViewRef) {
                    _this.embeddedViewRef.detectChanges();
                }
            })
        });
        this.embeddedViewRef = this.vcr.createEmbeddedView(this.templateRef, this.context);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    UseStateDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.useStateDefault) {
            this.value = this.useStateDefault;
        }
    };
    /**
     * @return {?}
     */
    UseStateDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.vcr.clear();
        this.embeddedViewRef.destroy();
        this.embeddedViewRef = null;
    };
    UseStateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[useState]'
                },] }
    ];
    /** @nocollapse */
    UseStateDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    UseStateDirective.propDecorators = {
        useStateDefault: [{ type: Input }]
    };
    return UseStateDirective;
}());
export { UseStateDirective };
if (false) {
    /** @type {?} */
    UseStateDirective.prototype.useStateDefault;
    /**
     * @type {?}
     * @private
     */
    UseStateDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    UseStateDirective.prototype.embeddedViewRef;
    /**
     * @type {?}
     * @private
     */
    UseStateDirective.prototype.value;
    /**
     * @type {?}
     * @private
     */
    UseStateDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    UseStateDirective.prototype.vcr;
}
