// @ngxf/platform/esm5/lib/directives/virtual.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/virtual.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import { VirtualHandler } from '../tools/virtual.handler';
var VirtualDirective = /** @class */ (function () {
    function VirtualDirective(handler, elementRef, cd) {
        this.handler = handler;
        this.elementRef = elementRef;
        this.cd = cd;
        this.handler.register(this.elementRef.nativeElement, cd);
    }
    /**
     * @return {?}
     */
    VirtualDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.handler.unregister(this.elementRef.nativeElement);
    };
    VirtualDirective.decorators = [
        { type: Directive, args: [{ selector: '[virtual]' },] }
    ];
    /** @nocollapse */
    VirtualDirective.ctorParameters = function () { return [
        { type: VirtualHandler },
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    VirtualDirective.propDecorators = {
        virtual: [{ type: Input }]
    };
    return VirtualDirective;
}());
export { VirtualDirective };
if (false) {
    /** @type {?} */
    VirtualDirective.prototype.virtual;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.handler;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.cd;
}
