// @ngxf/platform/esm2015/lib/tools/recompose/composed.view.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tools/recompose/composed.view.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { omit } from '../utils';
/** @type {?} */
export const PARENT_CONTEXT_TOKEN = Symbol('[PARENT_CONTEXT_TOKEN]');
/**
 * @record
 */
export function ComposedContext() { }
if (false) {
    /* Skipping unnamed member:
    [ PARENT_CONTEXT_TOKEN ]: TemplateRef<ComposedContext>[];*/
    /** @type {?} */
    ComposedContext.prototype.children;
    /* Skipping unhandled member: [ key: string ]: any;*/
}
/**
 * @template C
 */
export class ComposedView {
    /**
     * @param {?} viewContainerRef
     * @param {?=} templateRefs
     */
    constructor(viewContainerRef, templateRefs) {
        this.viewContainerRef = viewContainerRef;
        this.templateRefs = templateRefs;
    }
    /**
     * @param {?} context
     * @return {?}
     */
    createEmbeddedView(context) {
        const [templateRef, ...tail] = this.getParentTemplateRefs();
        this.context = (/** @type {?} */ ((/** @type {?} */ (context))));
        if (tail.length > 0) {
            this.elementRef = templateRef.elementRef;
            this.context[PARENT_CONTEXT_TOKEN] = tail;
            this.context.children = new ComposedView(this.viewContainerRef, tail);
        }
        this.viewRef =
            this.viewContainerRef.createEmbeddedView(templateRef, this.context);
        return (/** @type {?} */ ((/** @type {?} */ (this.viewRef))));
    }
    /**
     * @return {?}
     */
    markForCheck() {
        if (this.viewRef) {
            this.viewRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    detectChanges() {
        if (this.viewRef) {
            this.viewRef.detectChanges();
        }
    }
    /**
     * @param {?} update
     * @return {?}
     */
    updateContext(update) {
        /** @type {?} */
        const parent = this.getHydratedParentContext();
        update(this.context, parent);
        this.markForCheck();
    }
    /**
     * @return {?}
     */
    destroy() {
        if (this.viewRef && !this.viewRef.destroyed) {
            this.viewRef.destroy();
        }
        this.viewRef = null;
    }
    /**
     * @private
     * @return {?}
     */
    getParentTemplateRefs() {
        if (this.templateRefs) {
            return this.templateRefs;
        }
        /** @type {?} */
        const view = ((/** @type {?} */ (this.viewContainerRef.injector))).view;
        /** @type {?} */
        const context = findParentContext(view);
        return context[PARENT_CONTEXT_TOKEN];
    }
    /**
     * @private
     * @return {?}
     */
    getHydratedParentContext() {
        /** @type {?} */
        const view = ((/** @type {?} */ (this.viewContainerRef.injector))).view;
        return omit(findParentContext(view), [PARENT_CONTEXT_TOKEN, 'children']);
    }
}
if (false) {
    /** @type {?} */
    ComposedView.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    ComposedView.prototype.templateRefs;
}
/**
 * @param {?} view
 * @return {?}
 */
export function findParentContext(view) {
    /** @type {?} */
    const context = view.context;
    if (isParentContext(context)) {
        return context;
    }
    return findParentContext(view.parent);
}
/**
 * @param {?} context
 * @return {?}
 */
export function isParentContext(context) {
    return context && PARENT_CONTEXT_TOKEN in context;
}
