// @ngxf/platform/esm2015/lib/directives/is-dev-mode.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/is-dev-mode.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef, ViewContainerRef, isDevMode, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DevModeService {
    constructor() {
    }
    /**
     * @return {?}
     */
    isDevMode() {
        return isDevMode();
    }
}
DevModeService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
DevModeService.ctorParameters = () => [];
/** @nocollapse */ DevModeService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DevModeService_Factory() { return new DevModeService(); }, token: DevModeService, providedIn: "root" });
export class IsDevModeDirective {
    /**
     * @param {?} devModeService
     * @param {?} viewContainerRef
     * @param {?} templateRef
     */
    constructor(devModeService, viewContainerRef, templateRef) {
        if (devModeService.isDevMode()) {
            viewContainerRef.createEmbeddedView(templateRef);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
    }
}
IsDevModeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[isDevMode]'
            },] }
];
/** @nocollapse */
IsDevModeDirective.ctorParameters = () => [
    { type: DevModeService },
    { type: ViewContainerRef },
    { type: TemplateRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    IsDevModeDirective.prototype.viewRef;
}
