// @ngxf/platform/esm2015/lib/directives/repeat.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/repeat.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
export class RepeatContext {
}
if (false) {
    /** @type {?} */
    RepeatContext.prototype.$implicit;
    /** @type {?} */
    RepeatContext.prototype.repeat;
    /** @type {?} */
    RepeatContext.prototype.index;
}
export class RepeatDirective {
    /**
     * @param {?} viewContainerRef
     * @param {?} templateRef
     */
    constructor(viewContainerRef, templateRef) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
    }
    /**
     * @return {?}
     */
    get count() {
        /** @type {?} */
        const count = isNumber(this.repeat) ? this.repeat :
            isNumber(this.repeatOf) ? this.repeatOf : 0;
        return Math.max(count, 0);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if ('repeat' in changes || 'repeatOf' in changes) {
            this.onRepeatChanged(this.count);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.viewContainerRef.clear();
    }
    /**
     * @private
     * @param {?} count
     * @return {?}
     */
    onRepeatChanged(count) {
        /** When we need to create new items */
        for (let i = this.viewContainerRef.length; i < count; i++) {
            this.viewContainerRef.createEmbeddedView(this.templateRef, {
                $implicit: i, index: i, repeat: i
            }, i);
        }
        /** When we need to remove old items */
        for (let i = this.viewContainerRef.length; i > count; i--) {
            this.viewContainerRef.remove(i - 1);
        }
    }
}
RepeatDirective.decorators = [
    { type: Directive, args: [{ selector: '[repeat]' },] }
];
/** @nocollapse */
RepeatDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
RepeatDirective.propDecorators = {
    repeat: [{ type: Input }],
    repeatOf: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    RepeatDirective.prototype.repeat;
    /** @type {?} */
    RepeatDirective.prototype.repeatOf;
    /**
     * @type {?}
     * @private
     */
    RepeatDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    RepeatDirective.prototype.templateRef;
}
/**
 * @param {?} value
 * @return {?}
 */
function isNumber(value) {
    return typeof value === 'number';
}
