// @ngxf/platform/esm2015/lib/directives/return.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/return.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef } from '@angular/core';
import { ComposedView } from '../tools/recompose/composed.view';
/**
 * @record
 */
function ReturnContext() { }
if (false) {
    /** @type {?} */
    ReturnContext.prototype.$implicit;
}
export class ReturnDirective {
    /**
     * @param {?} viewContainerRef
     */
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if ('return' in changes) {
            if (!this.composedView) {
                this.create();
            }
            else {
                this.update();
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.composedView) {
            this.composedView.destroy();
            this.composedView = null;
        }
    }
    /**
     * @private
     * @return {?}
     */
    create() {
        this.composedView = new ComposedView(this.viewContainerRef);
        this.composedView.createEmbeddedView(this.context);
        this.update();
    }
    /**
     * @private
     * @return {?}
     */
    update() {
        this.composedView.updateContext((/**
         * @param {?} context
         * @return {?}
         */
        (context) => {
            return Object.assign(context, this.return);
        }));
    }
}
ReturnDirective.decorators = [
    { type: Directive, args: [{ selector: '[return]' },] }
];
/** @nocollapse */
ReturnDirective.ctorParameters = () => [
    { type: ViewContainerRef }
];
ReturnDirective.propDecorators = {
    return: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ReturnDirective.prototype.return;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.composedView;
    /**
     * @type {?}
     * @private
     */
    ReturnDirective.prototype.viewContainerRef;
}
