// @ngxf/platform/esm2015/lib/directives/timeout.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/timeout.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
export class TimeoutDirective {
    /**
     * @param {?} templateRef
     * @param {?} viewContainerRef
     */
    constructor(templateRef, viewContainerRef) {
        this.templateRef = templateRef;
        this.viewContainerRef = viewContainerRef;
    }
    /**
     * @param {?} milliseconds
     * @return {?}
     */
    set timeout(milliseconds) {
        this.dispose();
        if (isNumber(milliseconds)) {
            this.create(milliseconds);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.dispose();
        this.viewContainerRef.clear();
    }
    /**
     * @private
     * @param {?} milliseconds
     * @return {?}
     */
    create(milliseconds) {
        this.timeoutId = setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.viewContainerRef) {
                this.viewContainerRef.createEmbeddedView(this.templateRef);
            }
        }), milliseconds);
    }
    /**
     * @private
     * @return {?}
     */
    dispose() {
        this.viewContainerRef.clear();
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
    }
}
TimeoutDirective.decorators = [
    { type: Directive, args: [{ selector: '[timeout]' },] }
];
/** @nocollapse */
TimeoutDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef }
];
TimeoutDirective.propDecorators = {
    timeout: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.timeoutId;
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    TimeoutDirective.prototype.viewContainerRef;
}
/**
 * @param {?} value
 * @return {?}
 */
function isNumber(value) {
    return typeof value === 'number';
}
