// @ngxf/platform/esm2015/lib/directives/track-by-key.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/track-by-key.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgForOf } from '@angular/common';
import { Directive, Host, Input, Optional, ViewContainerRef } from '@angular/core';
/**
 * @template T
 */
export class NgForTrackByKeyDirective {
    /**
     * @param {?} ngFor
     * @param {?} viewContainerRef
     */
    constructor(ngFor, viewContainerRef) {
        this.ngFor = ngFor;
        this.viewContainerRef = viewContainerRef;
        if (!ngFor) {
            throw new Error('TrackByKey should use with *ngFor!');
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if ('ngForTrackByKey' in changes) {
            /** @type {?} */
            const key = this.ngForTrackByKey;
            if (key) {
                this.ngFor.ngForTrackBy = (/**
                 * @param {?} index
                 * @param {?} item
                 * @return {?}
                 */
                (index, item) => item[key]);
            }
            else {
                this.ngFor.ngForTrackBy = undefined;
            }
            this.ngFor['_differ'] = null;
            this.ngFor['_ngForOfDirty'] = true;
            this.viewContainerRef.clear();
            this.ngFor.ngDoCheck();
        }
    }
}
NgForTrackByKeyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngForTrackByKey]'
            },] }
];
/** @nocollapse */
NgForTrackByKeyDirective.ctorParameters = () => [
    { type: NgForOf, decorators: [{ type: Host }, { type: Optional }] },
    { type: ViewContainerRef }
];
NgForTrackByKeyDirective.propDecorators = {
    ngForTrackByKey: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgForTrackByKeyDirective.prototype.ngForTrackByKey;
    /**
     * @type {?}
     * @private
     */
    NgForTrackByKeyDirective.prototype.ngFor;
    /**
     * @type {?}
     * @private
     */
    NgForTrackByKeyDirective.prototype.viewContainerRef;
}
