// @ngxf/platform/esm2015/lib/directives/virtual.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/virtual.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import { VirtualHandler } from '../tools/virtual.handler';
export class VirtualDirective {
    /**
     * @param {?} handler
     * @param {?} elementRef
     * @param {?} cd
     */
    constructor(handler, elementRef, cd) {
        this.handler = handler;
        this.elementRef = elementRef;
        this.cd = cd;
        this.handler.register(this.elementRef.nativeElement, cd);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.handler.unregister(this.elementRef.nativeElement);
    }
}
VirtualDirective.decorators = [
    { type: Directive, args: [{ selector: '[virtual]' },] }
];
/** @nocollapse */
VirtualDirective.ctorParameters = () => [
    { type: VirtualHandler },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
VirtualDirective.propDecorators = {
    virtual: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    VirtualDirective.prototype.virtual;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.handler;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    VirtualDirective.prototype.cd;
}
