// @ngxf/socket.io/esm2015/lib/directives/socket.io.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/socket.io.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import * as io_ from 'socket.io-client';
/**
 * Fixes for rollup:
 * Error: Cannot call a namespace ('io')
 * @type {?}
 */
const io = io_;
/**
 * @record
 */
function SocketIOContext() { }
if (false) {
    /** @type {?} */
    SocketIOContext.prototype.$implicit;
}
export class SocketIODirective {
    /**
     * @param {?} templateRef
     * @param {?} vcr
     */
    constructor(templateRef, vcr) {
        this.templateRef = templateRef;
        this.vcr = vcr;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.socketIOOf && this.socketIOOf) {
            this.destroy();
            this.init();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroy();
    }
    /**
     * @private
     * @return {?}
     */
    init() {
        this.context.$implicit = this.socket = io(this.socketIOOf);
        this.vcr.createEmbeddedView(this.templateRef, this.context);
    }
    /**
     * @private
     * @return {?}
     */
    destroy() {
        this.vcr.clear();
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
        if (this.socket) {
            this.socket.disconnect();
            this.context.$implicit = this.socket = null;
        }
    }
}
SocketIODirective.decorators = [
    { type: Directive, args: [{
                selector: '[socketIO]'
            },] }
];
/** @nocollapse */
SocketIODirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef }
];
SocketIODirective.propDecorators = {
    socketIOOf: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SocketIODirective.prototype.socketIOOf;
    /**
     * @type {?}
     * @private
     */
    SocketIODirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    SocketIODirective.prototype.socket;
    /**
     * @type {?}
     * @private
     */
    SocketIODirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    SocketIODirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    SocketIODirective.prototype.vcr;
}
