// @ngxf/socket.io/esm2015/lib/directives/socket.io.on.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/socket.io.on.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { fromEvent } from 'rxjs';
import 'socket.io-client';
/**
 * @record
 */
function SocketIOOnContext() { }
if (false) {
    /** @type {?} */
    SocketIOOnContext.prototype.$implicit;
}
export class SocketIOOnDirective {
    /**
     * @param {?} templateRef
     * @param {?} vcr
     */
    constructor(templateRef, vcr) {
        this.templateRef = templateRef;
        this.vcr = vcr;
        this.context = {
            $implicit: null
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.socketIOOnOn || changes.socketIOOnFrom) {
            this.destroy();
            if (this.socketIOOnOn && this.socketIOOnFrom) {
                if (changes.socketIOOnOn.previousValue !== this.socketIOOnOn ||
                    changes.socketIOOnFrom.previousValue !== this.socketIOOnFrom) {
                    this.init();
                }
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroy();
    }
    /**
     * @private
     * @return {?}
     */
    init() {
        this.subscription = fromEvent(this.socketIOOnFrom, this.socketIOOnOn)
            .subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            this.context.$implicit = value;
            this.viewRef = this.templateRef.createEmbeddedView(this.context);
            this.viewRef.detectChanges();
            if (this.viewRef) {
                this.viewRef.destroy();
                this.viewRef = null;
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    destroy() {
        this.vcr.clear();
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
        if (this.viewRef) {
            this.viewRef.destroy();
            this.viewRef = null;
        }
    }
}
SocketIOOnDirective.decorators = [
    { type: Directive, args: [{
                selector: '[socketIOOn]'
            },] }
];
/** @nocollapse */
SocketIOOnDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef }
];
SocketIOOnDirective.propDecorators = {
    socketIOOnOn: [{ type: Input }],
    socketIOOnFrom: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SocketIOOnDirective.prototype.socketIOOnOn;
    /** @type {?} */
    SocketIOOnDirective.prototype.socketIOOnFrom;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    SocketIOOnDirective.prototype.vcr;
}
